### ISSUE TITLE
Add toString method to vanilla Tag class

### ISSUE THREAD

**Issue by mcenderdragon** [MinecraftForge/MinecraftForge#7570]
Because this error is not verbose is suggest adding some information to the Tag instance:

```
java.lang.IllegalStateException: Duplicate key net.minecraft.tags.Tag@35be503c
	at java.util.stream.Collectors.lambda$throwingMerger$0(Unknown Source) ~[?:1.8.0_211] {}
	at java.util.HashMap.merge(Unknown Source) ~[?:1.8.0_211] {}
	at java.util.stream.Collectors.lambda$toMap$58(Unknown Source) ~[?:1.8.0_211] {}
	at java.util.stream.ReduceOps$3ReducingSink.accept(Unknown Source) ~[?:1.8.0_211] {}
	at java.util.stream.ReferencePipeline$2$1.accept(Unknown Source) ~[?:1.8.0_211] {}
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source) ~[?:1.8.0_211] {}
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:1.8.0_211] {}
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:1.8.0_211] {}
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source) ~[?:1.8.0_211] {}
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:1.8.0_211] {}
	at java.util.stream.ReferencePipeline.collect(Unknown Source) ~[?:1.8.0_211] {}
	at net.minecraft.tags.TagRegistry.reinjectOptionalTags(TagRegistry.java:89) ~[forge-1.16.4-35.1.7_mapped_snapshot_20201028-1.16.3-recomp.jar:?] {re:classloading}
	at net.minecraft.tags.ITagCollectionSupplier.reinjectOptionalTags(ITagCollectionSupplier.java:66) ~[forge-1.16.4-35.1.7_mapped_snapshot_20201028-1.16.3-recomp.jar:?] {re:classloading}
```

**Comment by TheCurle** [comment]
Do you have a minimal test case to reproduce this error for testing?

**Comment by apetschenka** [comment]
please assign this to me, would like to work on this .

**Comment by AterAnimAvis** [comment]
Hello 👋, whilst we do welcome contributions, we don’t assign people to issues like that.

Please note there’s already an open PR to fix this in #8112. 

As it looks like you may be new, please ensure you check out the Contribution guides linked in the README, Forge works slightly differently from normal projects.

**Comment by sciwhiz12** [comment]
So, I've investigated this issue a bit more and the underlying causes while looking at #8112 as a potential solution, and I've found a way to reproduce an error which leads to the same stacktrace as the issue provides, along with my potential solution for the problem.

### Reproducing the issue

To reproduce the error:
 - Insert the following line twice somewhere:
   ```java
   public static final Tags.IOptionalNamedTag<Block> TESTY_TAG = BlockTags.createOptional(new ResourceLocation("testy:test_test"));
   ```
   The location does not matter, though this would likely be found in an actual end-user scenario across two mods which define the same option tag. Neither the default contents of the tag, the actual type of objects in the tag, nor the location/name of the tag matter for this; however, this only occurs if two tags of the exact same object type and location/name are registered twice or more.
 - _Observe a datapack loading error upon loading any world, and a similar stacktrace to the issue's stacktrace in your console/logs._

### Explanation

#### How missing optional tags are handled

The underlying problem is how `StaticTagHandler` (or `TagRegistry` in 1.16) handles reinjecting optional tags -- that is, adding the default values of optional tags if they are not present in the main collection of defined tags.

Each invocation of `BlockTags.createOptional` creates an `OptionalNamedTag` object, which is registered at the `StaticTagHandler` for the specific object type of the tag. These objects when registered are added to a `wrappers` list within the tag handler.

When reinjecting optional tags into a `TagCollection` of the same type, it follows the following process:
 - Retrieve all tags within the tag collection, which is represented by a `Map<ResourceLocation, Tag>`.
 - Stream the contents of the `wrappers` list, filtering for those which are instances of `OptionalNamedTag` (in other words, optional tags) and whose name does not appear in the map of all tags within the tag collection above. These missing optional tags are then collected into a `Map<ResourceLocation, Tag>`, where the key is the name of the optional tag and the value is calculated by first marking the optional tag as 'defaulted' and then resolving the default values into a `Tag`.
 - If there are no missing optional tags, return the tag collection unchanged. However, if there _are_ missing optional tags, add the map of tags from the original tag collection to the map of defaulted optional tags, and construct and return a new tag collection from that combined map.

#### How it explodes

The problem lies with the use of `Collectors.toMap(Function, Function)` when collecting the defaulted optional tags map. This collector, according to its javadoc, has the following property (the second sentence will be of importance later):

> If the mapped keys contain duplicates (according to `Object.equals(Object)`), an `IllegalStateException` is thrown when the collection operation is performed. If the mapped keys might have duplicates, use toMap(Function, Function, BinaryOperator) instead. 
> ~ [`Collectors#toMap(Function, Function)`, Java SE 16 and JDK 16](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/stream/Collectors.html#toMap(java.util.function.Function,java.util.function.Function))

This presents a problem when two or more `OptionalNamedTag` wrapper objects exist in the `wrappers` list (due to being registered twice or more, such as by two separate mods) point to the same optional tag. Let us take a sample step-by-step:

 - The first tag wrapper is collected, mapped into its tag name and defaulted values as explained above, and inserted into the map by the collector as normal. As no tags of the same name have been processed beforehand, this is fine.
 - The second tag wrapper is collected, mapped into its tag name _(which is the same as the first one's)_ and defaulted values, and inserted into the map by the collector.
    At this point, the collector notices that the key -- the tag name -- already exists within the map due to the first tag wrapper, and promptly throws an `IllegalStateException` to complain of this.

This means that if an optional tag is defined in code at least twice and it is missing, an error will be raised due to the duplicate tag names within the map, as demonstrated by the reproduction code.

### Proposed solution

My proposed solution here is to take heed of the advice mentioned in the above javadoc: to use the other overload of `Collectors#toMap` which takes in a merging function, and supply an appropriate function. This function will likely be one that combines the values of both tags into a single tag. 

This does pose a potential problem if different mods supply different defaulted values for the same optional tag, as they will be combined and therefore the mod features which use this tag may have unexpected behaviors with objects not originally intended by the mod authors. A possible alternative merging function would be to return an empty tag if the defaulted values from both wrappers are not the same.

Inserting the following as the merging function for the `toMap` call within `StaticTagHandler#reinjectOptionalTags` fixes the above error in the reproduction steps:
```java
(t1, t2) -> SetTag.create(com.google.common.collect.ImmutableSet.<T>builder().addAll(t1.getValues()).addAll(t2.getValues()).build())
```
